### PR TITLE
Stage 5 Remaster

### PR DESCRIPTION
**Pull request by StarWitch**
Modernizing Stage 5. WIP so commits can be commented on.

### CODE REVIEW

File: src/stages/stage5/timeline.c [hunk 1]
```diff
@@ -400,17 +366,114 @@ static int stage5_superbullet(Enemy *e, int t) {
...
+}) {
+	Enemy *e = TASK_BIND(ARGS.e);
+	cmplx dir = cnormalize(global.plr.pos - e->pos);
+
+	for(int i = -ARGS.count; i <= ARGS.count; i++) {
+		PROJECTILE(
+			.proto = pp_bullet,
+			.pos = e->pos,
+			.color = RGB(0.0, 0.0, 1.0),
+			.move = move_asymptotic_simple(3.5 * cdir(carg(dir) + 0.06 * i), 5),
```
Comment (Akaricchi) [inline src/stages/stage5/timeline.c:381]: It's generally more efficient and cleaner to do `dir * cdir(X)` instead of `cdir(carg(dir) + X)`.

Also in the original code, the bullet speed is 4.5 on Lunatic (that's the `+(global.diff == D_Lunatic)` term). I suggest defining a `real speed = difficulty_value(3.5, 3.5, 3.5, 4.5);` outside the loop

File: src/stages/stage5/timeline.c [hunk 2]
```diff
@@ -400,17 +366,114 @@ static int stage5_superbullet(Enemy *e, int t) {
...
+TASK(greeter_fairy, {
+		cmplx pos;
+		MoveParams move_enter;
+		MoveParams move_exit;
+}) {
+	Enemy *e = TASK_BIND(espawn_fairy_blue_box(ARGS.pos, ITEMS(.points = 1)));
+	e->move = ARGS.move_enter;
+
+	for (int i = 0; i < 5; i++) {
+		INVOKE_SUBTASK_DELAYED(80, greeter_fairy_attack, {
```
Comment (Akaricchi) [inline src/stages/stage5/timeline.c:396]: I think you could rewrite this into a normal function instead of a task. It doesn't have any async control flow, and the delayed invocation can be replaced with just a `WAIT(80)` before the loop. The `WAIT(200)` in the end would have to be shortened to a `WAIT(20)` to match the original behavior.

With delayed invocations it doesn't match the original either by the way, because the loop already waits for `20 * 5 = 100` frames.

Either of those should match:

```c
WAIT(80);
for(int i = 0; i < 5; i++) {
	greeter_fairy_attack(...);
	WAIT(20);
}
WAIT(20);
e->move = ARGS.move_exit;
```

```c
for(int i = 0; i < 5; i++) {
	INVOKE_SUBTASK_DELAYED(80, greeter_fairy_attack, ...);
	WAIT(20);
}
WAIT(100);
e->move = ARGS.move_exit;
```

```c
for(int i = 0; i < 5; i++) {
	INVOKE_SUBTASK_DELAYED(80 + i * 5, greeter_fairy_attack, ...);
}
WAIT(200);
e->move = ARGS.move_exit;
```

I would personally choose number 1 because, as I said, the subtask is not really necessary here. You could also just inline the "attack" function.

File: src/stages/stage5/timeline.c [hunk 3]
```diff
@@ -400,17 +366,114 @@ static int stage5_superbullet(Enemy *e, int t) {
...
+			.count = difficulty_value(1, 2, 3, 4),
+		});
+		WAIT(20);
+	}
+
+	WAIT(200);
+
+	e->move = ARGS.move_exit;
+
+	STALL;
```
Comment (Akaricchi) [inline src/stages/stage5/timeline.c:407]: The `STALL` is not needed in this case. It's generally used to keep subtasks alive, but:
* All subtasks invoked here are short enough that they'll surely be finished before this code is reached;
* If you implement my earlier suggestion, there wouldn't be subtasks at all.

**Review by Akaricchi — COMMENTED** [review]
Looks mostly fine. Although it's not required to port it perfectly, I left some suggestions on how this could be made a bit more faithful to the original, just to help you sharpen your intuition.

File: src/stages/stage5/timeline.c [hunk 4]
```diff
@@ -334,197 +135,713 @@ static int stage5_magnetto(Enemy *e, int t) {
...
-	AT(2500) {
-		create_enemy1c(VIEWPORT_W/2, 2000, BigFairy, stage5_lightburst, 2.0*I);
+TASK(lightburst_fairies_1, {
+	int num;
+	cmplx pos1;
+	cmplx pos2;
+	cmplx exit;
+}) {
+	for (int i = 0; i < ARGS.num; i++) {
+		cmplx pos = ARGS.pos1 + ARGS.pos2 * i;
```
Comment (Akaricchi) [inline src/stages/stage5/timeline.c:344]: Those parameter names seem misleading. You'd think `pos1` and `pos2` are both absolute positions, but in fact `pos2` seems to be used like an offset to `pos1`.

File: src/stages/stage5/timeline.c [hunk 5]
```diff
@@ -334,197 +135,713 @@ static int stage5_magnetto(Enemy *e, int t) {
...
+		WAIT(40);
 	}
+}
 
-	AT(2900)
-		global.boss = stage5_spawn_midboss();
+TASK(loop_swirl_move, {
+    MoveParams *move;
+	cmplx turn;
+	cmplx retention;
```
Comment (Akaricchi) [inline src/stages/stage5/timeline.c:374]: Bad indent

File: src/stages/stage5/timeline.c [hunk 6]
```diff
@@ -334,197 +135,713 @@ static int stage5_magnetto(Enemy *e, int t) {
...
+}) {
+	int begin = creal(ARGS.turn) + 1;
+	int end = cimag(ARGS.turn) - 1;
+	int time = end - begin;
+
+	ARGS.move->retention = ARGS.retention;
+	WAIT(time / 1.5);
```
Comment (Akaricchi) [inline src/stages/stage5/timeline.c:381]: Why `/ 1.5`?

File: src/stages/stage5/timeline.c [hunk 7]
```diff
@@ -334,197 +135,713 @@ static int stage5_magnetto(Enemy *e, int t) {
...
-		// TODO: determine if this hack is still required with the new dialogue system
-		global.timer++;
+    e->move = move_linear(ARGS.velocity);
+    INVOKE_SUBTASK_DELAYED(80, loop_swirl_move,
+        .move = &e->move,
+		.turn = ARGS.turn,
+		.retention = ARGS.retention
+    );
+
+	int difficulty = 30 - difficulty_value(4, 8, 16, 24);
```
Comment (Akaricchi) [inline src/stages/stage5/timeline.c:400]: Broken indents

File: src/stages/stage5/timeline.c [hunk 8]
```diff
@@ -334,197 +135,713 @@ static int stage5_magnetto(Enemy *e, int t) {
...
+		for(int i = 1; i >= -1; i -= 2) {
+			PROJECTILE(
+				.proto = pp_bullet,
+				.pos = e->pos,
+				.color = RGB(0.3, 0.4, 0.5),
+				.move = move_asymptotic_simple(i * 2 * ARGS.velocity * I / cabs(ARGS.velocity), 3),
+				// TODO: bullets only fire down-left with the above
+				// they should also fire in other directions correlating to the angle of "turn" the swirls make
+				//.rule = asymptotic,
+				//.args = { i*2*e->args[0]*I/cabs(e->args[0]), 3 }
```
Comment (Akaricchi) [inline src/stages/stage5/timeline.c:412]: Sounds like you want to use `e->move.velocity` instead of `ARGS.velocity` then.

Also, you have a `velocity / cabs(velocity)` there, which should be simplified to `cnormalize(velocity)`

File: src/stages/stage5/timeline.c [hunk 9]
```diff
@@ -334,197 +135,713 @@ static int stage5_magnetto(Enemy *e, int t) {
...
+
+	int difficulty = difficulty_value(0.25, 0.50, 0.75, 1);
+	for (int x = 0; x < 400; x++) {
+		double base_angle = carg(global.plr.pos - e->pos);
 
-	FROM_TO(3300, 4000, 90-10*global.diff)
-		create_enemy1c(200.0*I+VIEWPORT_W*(_i&1), 1500, Fairy, stage5_superbullet, 3-6*(_i&1));
+		for (int i = 1; i >= -1; i -= 2) {
+			double a = i * 0.2 - 0.1 * difficulty + i * 3.0 / (x + 1);
+			cmplx aim = cdir(base_angle + a);
```
Comment (Akaricchi) [inline src/stages/stage5/timeline.c:453]: 1. Don't use `double`, use `real`
2. Better store the aiming direction vector e.g. `cmplx plraim = cnormalize(global.plr.pos - e->pos)`. Then you can rotate it by an arbitrary angle with `cmplx aim = plraim * cdir(a)`

File: src/stages/stage5/timeline.c [hunk 10]
```diff
@@ -334,197 +135,713 @@ static int stage5_magnetto(Enemy *e, int t) {
...
-		create_enemy1c(VIEWPORT_W/2, 2000, BigFairy, stage5_lightburst, 2.0*I);
+	int difficulty = difficulty_value(8, 6, 4, 3);
+
+	for (int x = 0; x < 600; x++) {
+		PROJECTILE(
+			.proto = pp_rice,
+			.pos = e->pos + 20 * cdir(2.0 * M_PI * rng_real()),
+			.color = RGB(0, 0, 255),
+			.move = move_linear(cdir(2.0 * M_PI * rng_real())),
+		);
```
Comment (Akaricchi) [inline src/stages/stage5/timeline.c:501]: Do not put multiple RNG calls in the same statement. The evaluation order is undefined and may break replays across builds (because C is awesome like that). We had the `a*rand*` facility (and its successor, `RNG_ARRAY` and `vrng_*`) to hack around this problem, but it's usually better to just use variables.

`ninja upkeep` can detect problematic RNG usage, although it has false positives.

Note that this is also sometimes neglected in effect-spawning code, because it ultimately has no effect on game state, but that's still not a good practice.

**Comment by Akaricchi** [comment]
Github was "helpful" enough to hide some of my comments, beware.

File: src/stages/stage5/timeline.c (hunk 6 above)
Comment (StarWitch) [reply src/stages/stage5/timeline.c:381]: It's the only way to get the swirls to end their "looping" motion at the correct angle. I'm not too sure why, as I was mostly roughing it out at the time.

Comment (Akaricchi) [reply src/stages/stage5/timeline.c:385]: That's not the right way to go about fixing it. You should either change the rotation angle-per-frame (the argument to `cdir()` used for `retention`), or adjust the begin/end timing at call site. There are also a couple more problems with this code that I haven't noticed at first:

* First of all there's no reason to express the timings as a single `cmplx` parameter. The only reason the old code does it is because in the crappy old system you could only have `cmplx` args (and only 4 of them at that). You should just make `int end, begin` task parameters; I'd name them `turn_begin/turn_end` for clarity.
* The +/-1 offsets should be just moved to the call site.
* The `begin` timing isn't used properly. You should `WAIT(begin)` before changing retention to a rotating one. And then you shouldn't use `INVOKE_SUBTASK_DELAYED` - set the `begin` timing properly instead.
* Does the `WAIT()` even work? It seems like you're passing a negative number into it:
    * You are calling the `loop_swirls` task with `turn = 9`. This is looks very wrong...
    * `begin == creal(9) + 1 == 9 + 1 == 10`
    * `end == cimag(9) - 1 == 0 - 1 == -1`
    * `time == end - begin == -1 - 10 == -11`
    * `WAIT(time / 1.5) = WAIT(-7)` - honestly not even sure what happens in this case. We probably should have an assertion for wait time >= 0 inside that macro.


Once you get this all figured out (and preferably *not* before you do), I'd recommend switching to a more intuitive set of outermost parameters for turning:
* `int turn_delay` - time to wait before initiating the turn, in frames.
* `int turn_duration` - duration of the turn, in frames.
* `real turn_angle` - total angle of the turn, in radians.

This is exactly what the stage 2 fairy turning code does, and perhaps it makes sense to generalize and move it to `common_tasks.c` eventually.

Comment (StarWitch) [reply src/stages/stage5/timeline.c]: Done and done. I was worried the movement was going to be to different but wiping out the old format and using what was in Stage 2 ended up being much better. I also moved it to `common_tasks`. I did a before/after comparison and the movement in Stage 2 seemed identical, so there you go.

File: src/stages/stage5/timeline.c [hunk 11]
```diff
@@ -37,537 +27,854 @@ static void stage5_dialog_post_midboss(void) {
...
-	}
+	e->move = move_towards(e->pos + ARGS.acceleration * 70 + ARGS.offset, 0.05);
+	WAIT(60);
 
-	AT(6960) {
-		stage_unlock_bgm("stage5");
-		global.boss = stage5_spawn_boss();
+	int difficulty = difficulty_value(1, 2, 3, 4);
+	for (int x = 0; x < 140; x++) {
+		cmplx n = cdir(M_PI * sin(x / (8.0 + difficulty) + rng_real() * 0.1) * carg(global.plr.pos - e->pos));
```
Comment (Akaricchi) [inline src/stages/stage5/timeline.c:594]: Looks suspiciously wrong. The original line was this, I think?
```c
cmplx n = cexp(I*M_PI*sin(_i/(8.0+global.diff)+frand()*0.1)+I*carg(global.plr.pos-e->pos));
```

In that case it is almost correct, except instead of multiplying by the `carg(...)` you have to add it, like the original does... and then you preferably get rid of the `carg`. I explain that part in detail below.

The expression is confusing so let's break it down.

**Step 0**: de-laochainlanize the code:
```c
cmplx n = cexp(I * M_PI * sin(_i / (8.0 + global.diff) + frand() * 0.1) + I * carg(global.plr.pos - e->pos));
```

**Step 1**: replace `cexp(I * x)` with `cdir(x)`. There are 2 `I`s here, so let's factor them out:
```c
cmplx n = cexp(I * (M_PI * sin(_i / (8.0 + global.diff) + frand() * 0.1) + carg(global.plr.pos - e->pos)));
```
Then we can substitute `cdir()` easily:
```c
cmplx n = cdir(M_PI * sin(_i / (8.0 + global.diff) + frand() * 0.1) + carg(global.plr.pos - e->pos));
```

**Step 2**: get rid of that `carg`:

What we have here is a sum of two angles, `A` and `B`
```c
real A = M_PI * sin(_i / (8.0 + global.diff) + frand() * 0.1);
real B = carg(global.plr.pos - e->pos);
cmplx n = cdir(A + B);
```

Since complex multiplication is equivalent to a scale and a rotation, we can rewrite `n` as follows:
```c
cmplx n = cdir(A) * cdir(B);
```

If we represent A and B as direction vectors rather than angles, we get:
```c
cmplx A = cdir(M_PI * sin(_i / (8.0 + global.diff) + frand() * 0.1));
cmplx B = cdir(carg(global.plr.pos - e->pos));
cmplx n = A * B;
```

Let's focus on B now:
```c
cdir(carg(global.plr.pos - e->pos))
```
This expression basically says "get the angle (i.e. argument) of complex `global.plr.pos - e->pos`, and make another complex with the same angle but magnitude = 1". That's a very roundabout way of spelling `cnormalize(global.plr.pos - e->pos)`.
It's also mathematically equivalent to `(global.plr.pos - e->pos) / cabs(global.plr.pos - e->pos)` (`cnormalize` does a slightly optimized version of that).

So:
```c
cmplx A = cdir(M_PI * sin(_i / (8.0 + global.diff) + frand() * 0.1));
cmplx B = cnormalize(global.plr.pos - e->pos);
cmplx n = A * B;
```

Folding A and B back into n:
```c
cmplx n = cdir(M_PI * sin(_i / (8.0 + global.diff) + frand() * 0.1)) * cnormalize(global.plr.pos - e->pos);
```

### TL;DR

To deal with such expressions you just need to know the following identities:

```
cdir(A + B)  ==  cdir(A) * cdir(B)
```
```
cdir(carg(X))  ==  cnormalize(X)
```
And from these follows:
```
cdir(A + carg(B))  ==  cdir(A) * cnormalize(B)
```

**Review by Akaricchi — COMMENTED** [review]
Not a super comprehensive review, admittedly I skimmed through some parts so apologies if I missed anything important; I'll come back to it later. Also github is probably going to hide some of my comments again.

**Comment by StarWitch** [comment]
One thing to note is that some of the `timeline.c` enemy timings need to be tweaked a bit. I tried to replicate them 1:1 with the old system but some appear too soon/too late for various reasons. I'll work on that once `Overload` is finished.

File: src/stages/stage5/spells/artificial_lightning.c [hunk 12]
```diff
@@ -10,148 +10,155 @@
...
+	Enemy *e = TASK_BIND(create_enemy1c(ARGS.pos, ENEMY_IMMUNE, NULL, NULL, 0));
 
-	FROM_TO(0,200,20)
-		e->args[0] *= cexp(I * (0.25 + 0.25 * frand() * M_PI));
+	INVOKE_TASK(lightning_slave_move, ENT_BOX(e), ARGS.move_arg);
 
-	FROM_TO(0, 200, 3)
-		if(cabs(e->pos-global.plr.pos) > 60) {
-			Color *clr = RGBA(1-1/(1+0.01*_i), 0.5-0.01*_i, 1, 0);
+	for(int x = 0; x < 67; x += WAIT(3)) {
```
Comment (laochailan) [inline src/stages/stage5/spells/artificial_lightning.c:87]: what is `x += WAIT(3)`? i am not a fan of these strange for-loop reinvention constructions. but if it has to be done, `x++, WAIT(3)` is better

File: src/stages/stage5/spells/artificial_lightning.c [hunk 13]
```diff
@@ -10,148 +10,155 @@
...
+				// TODO: this doesn't quite work right
+				.move = move_asymptotic_simple(0.75 * ARGS.move_arg / cabs(ARGS.move_arg) * I, 10),
 			);
 
 			if(projectile_in_viewport(p)) {
 				for(int i = 0; i < 3; ++i) {
-					tsrand_fill(2);
-					iku_lightning_particle(p->pos + 5 * afrand(0) * cexp(I*M_PI*2*afrand(1)), 0);
+					RNG_ARRAY(rand, 2);
+					iku_lightning_particle(p->pos + 5 * vrng_sreal(rand[0]) * cdir(M_TAU * vrng_sreal(rand[1])));
```
Comment (laochailan) [inline src/stages/stage5/spells/artificial_lightning.c:102]: `vrng_sreal` -> `vrng_real` although here it doesn’t affect the result.

File: src/stages/stage5/spells/atmospheric_discharge.c [hunk 14]
```diff
@@ -10,61 +10,61 @@
...
-		cmplx p2 = p1 + (120+20*global.diff)*cexp(0.5*I-afrand(2)*I)*(1-2*(afrand(3) > 0.5));
-
-		int i;
-		int c = 6+global.diff;
+TASK(ball_shoot, NO_ARGS) {
+	int difficulty = difficulty_value(1, 2, 3, 4);
+	for(int x = 0;; x += WAIT(difficulty_value(19, 17, 15, 13))) {
+		RNG_ARRAY(rand, 4);
+		cmplx p1 = VIEWPORT_W * vrng_real(rand[0]) + VIEWPORT_H/2 * I * vrng_real(rand[1]);
+		cmplx p2 = p1 + (120 + 20 * difficulty) * cdir(0.5 - vrng_real(rand[2])) * (1 - 2 * (vrng_real(rand[3]) > 0.5));
```
Comment (laochailan) [inline src/stages/stage5/spells/atmospheric_discharge.c:38]: `+ difficulty_value(140, 160, 180, 200)` or `+20*difficulty_value(7,8,9,10)`

`cdir(0.5 - vrng_real(rand[2]))` -> `cdir(0.5*vrng_sreal(rand[2]))`

for the last one maybe `vrng_chance(rand[3], 0.5)` is better?

File: src/stages/stage5/spells/atmospheric_discharge.c [hunk 15]
```diff
@@ -10,61 +10,61 @@
...
 
-		for(i = -c*0.5; i <= c*0.5; i++) {
+		int count = difficulty_value(7, 8, 9, 10);
```
Comment (laochailan) [inline src/stages/stage5/spells/atmospheric_discharge.c:40]: this should be 1,2,3,4 as it replaces global.diff below!

File: src/stages/stage5/spells/atmospheric_discharge.c [hunk 16]
```diff
@@ -10,61 +10,61 @@
...
 				.proto = pp_ball,
-				.pos = p1+(p2-p1)/c*i,
-				.color = RGBA(1-1/(1+fabs(0.1*i)), 0.5-0.1*abs(i), 1, 0),
-				.rule = accelerated,
-				.args = {
-					0, (0.004+0.001*global.diff)*cexp(I*carg(p2-p1)+I*M_PI/2+0.2*I*i)
-				},
+				.pos = p1 + (p2 - p1) / count * i,
+				.color = RGBA(1 - 1 / (1 + fabs(0.1 * i)), 0.5 - 0.1 * abs(i), 1, 0),
+				.move = move_accelerated(0, (0.004 + 0.001 * difficulty) * cnormalize(p2 - p1) * cdir(M_PI/2 + 0.2 * i)),
```
Comment (laochailan) [inline src/stages/stage5/spells/atmospheric_discharge.c:46]: better yet:
`0.001*difficulty_value(5,6,7,8)`

File: src/stages/stage5/timeline.c [hunk 17]
```diff
@@ -37,537 +26,845 @@ static void stage5_dialog_post_midboss(void) {
...
-			cmplx n = cexp(I*carg(global.plr.pos) + 2.0*I*M_PI/c*i);
+TASK(magnetto_swirl_shoot, {
+	BoxedEnemy e;
+	cmplx move_to;
+}) {
+	Enemy *e = TASK_BIND(ARGS.e);
+
+	int interval = difficulty_value(12, 9, 6, 3);
+	int bullets = difficulty_value(1, 2, 2, 2);
+	int bullet_dir = difficulty_value(0, 0, 1, 0);
```
Comment (laochailan) [inline src/stages/stage5/timeline.c:39]: why is a direction valued 0 and 1?

File: src/stages/stage5/timeline.c [hunk 18]
```diff
@@ -37,537 +26,845 @@ static void stage5_dialog_post_midboss(void) {
...
-				.pos = e->pos + 50*n*cexp(-1.0*I*_i*global.diff),
-				.color = RGB(0.3, 0, 0.7+0.3*(_i&1)),
-				.rule = asymptotic,
-				.args = {
-					2.5*n+0.25*global.diff*afrand(0)*cexp(2.0*I*M_PI*afrand(1)),
-					3
-				}
+				.pos = e->pos + 50 * n * cdir(-1.0 * x * difficulty),
+				.color = RGB(0.3, 0, 0.7 + 0.3 * (i % 2)),
+				.move = move_asymptotic_simple(2.5 * n + 0.25 * difficulty * rng_sreal() * rng_dir(), 3),
```
Comment (laochailan) [inline src/stages/stage5/timeline.c:278]: RNG POLICE, OPEN UP!!!

File: src/stages/stage5/timeline.c [hunk 19]
```diff
@@ -37,537 +26,845 @@ static void stage5_dialog_post_midboss(void) {
...
-		create_enemy1c(VIEWPORT_W*(_i&1)+70.0*I+50*_i*I, 400, Fairy, stage5_greeter, 3-6*(_i&1));
-	}
+	int delay = difficulty_value(6, 5, 4, 3) + ARGS.reduction;
+	int amount = 700 / delay;
+	int difficulty = difficulty_value(1, 2, 3, 4);
 
-	FROM_TO(270, 320, 40)
-		create_enemy1c(VIEWPORT_W/4+VIEWPORT_W/2*_i, 2000, BigFairy, stage5_lightburst, 2.0*I);
+	for(int x = 0; x < amount; x++) {
+		cmplx n = cnormalize(global.plr.pos - e->pos) * cdir((0.2 - 0.02 * difficulty) * x);
```
Comment (laochailan) [inline src/stages/stage5/timeline.c:368]: `0.2 - 0.02 * difficulty`
-> `0.02 * difficulty_value(9,8,7,6)`

File: src/stages/stage5/timeline.c [hunk 20]
```diff
@@ -37,537 +26,845 @@ static void stage5_dialog_post_midboss(void) {
...
+	e->move = move_linear(ARGS.velocity);
 
-	AT(5360) {
-		create_enemy1c(30.0*I+VIEWPORT_W, 1500, Fairy, stage5_superbullet, -2 + I);
+	int difficulty = difficulty_value(8, 6, 4, 3);
 
+	for(int x = 0; x < 600; x++) {
+		PROJECTILE(
+			.proto = pp_rice,
+			.pos = e->pos + 20 * cdir(M_TAU * rng_real()),
```
Comment (laochailan) [inline src/stages/stage5/timeline.c:519]: rng_dir

File: src/stages/stage5/timeline.c [hunk 21]
```diff
@@ -37,537 +26,845 @@ static void stage5_dialog_post_midboss(void) {
...
+			.color = RGB(0, 0, 255),
+			.move = move_linear(cdir(M_TAU  * rng_real())),
```
Comment (laochailan) [inline src/stages/stage5/timeline.c:521]: rng_dir

**Review by laochailan — COMMENTED** [review]
I’m not sure about replacing every `global.diff` by `difficulty = difficulty_value(1,2,3,4)` if no further simplifications are possible. I tried to do the simplifications where it came up in my comments.

And I probably missed stuff some task-related subtleties!

File: src/stages/stage5/spells/artificial_lightning.c [hunk 22]
```diff
@@ -10,148 +10,155 @@
...
+	for(int time = 0;; time++, YIELD) {
+		if(time > 0 && !(time % 100)) {
+			int count = difficulty_value(7, 7, 7, 9);
+			for(int i = 0; i < count; i++) {
+				Projectile *p = PROJECTILE(
+					.proto = pp_bigball,
+					.pos = boss->pos,
+					.color = RGBA(0.5, 0.1, 1.0, 0.0),
+				);
+				INVOKE_SUBTASK(zigzag_move, ENT_BOX(p), cdir(M_TAU / count * i) * cnormalize(global.plr.pos - boss->pos));
```
[the remark this answers is not in the document]
Comment (Akaricchi) [reply src/stages/stage5/spells/artificial_lightning.c:61]: No it shouldn't be, it already binds to the projectile lifetime which should be enough.

File: src/stages/stage5/spells/artificial_lightning.c [hunk 23]
```diff
@@ -10,148 +10,155 @@
...
-static int lightning_slave(Enemy *e, int t) {
-	if(t < 0)
-		return 1;
-	if(t > 200)
-		return ACTION_DESTROY;
+TASK(zigzag_shoot, { BoxedBoss boss; } ) {
+	Boss *boss = TASK_BIND(ARGS.boss);
+	for(int time = 0;; time++, YIELD) {
+		if(time > 0 && !(time % 100)) {
+			int count = difficulty_value(7, 7, 7, 9);
```
Comment (Akaricchi) [inline src/stages/stage5/spells/artificial_lightning.c:54]: I think you want something like this instead

```c
for(int time = 0;;) {
	time += WAIT(delay);
	int count = difficulty_value(7, 7, 7, 9);
```

File: src/stages/stage5/spells/artificial_lightning.c (hunk 12 above)
Comment (Akaricchi) [reply src/stages/stage5/spells/artificial_lightning.c:87]: @laochailan that's not equivalent. `WAIT()` returns the amount of frames waited, so you'd have to do `x += 3, WAIT(3)` if you really wanted to rewrite this. But that's repeating yourself, and not any better in my opinion.

File: src/stages/stage5/spells/artificial_lightning.c (hunk 13 above)
Comment (Akaricchi) [reply src/stages/stage5/spells/artificial_lightning.c:102]: Actually just replace that whole `cdir(M_TAU * vrng_sreal(rand[1])))` with `vrng_dir(rand[1])`.

File: src/stages/stage5/spells/atmospheric_discharge.c (hunk 15 above)
Comment (Akaricchi) [reply src/stages/stage5/spells/atmospheric_discharge.c:40]: Doesn't seem to be the case? `int difficulty = difficulty_value(1, 2, 3, 4);` at the top of the task replaces `global.diff` (not very useful in my opinion, but correct). The `count` here replaces `int c = 6+global.diff;`, which seems fine. Am I missing something?

File: src/stages/stage5/spells/atmospheric_discharge.c (hunk 16 above)
Comment (Akaricchi) [reply src/stages/stage5/spells/atmospheric_discharge.c:46]: Also in general I would prefer to have all the constant `difficulty_value(...)` expressions grouped up in variables outside of all the loops, preferably with meaningful names, for easier tuning and general readability.

File: src/stages/stage5/spells/artificial_lightning.c (hunk 12 above)
Comment (laochailan) [reply src/stages/stage5/spells/artificial_lightning.c:87]: then this is wrong because the amount of iterations has been shortened to 200/3.

File: src/stages/stage5/spells/atmospheric_discharge.c (hunk 15 above)
Comment (laochailan) [reply src/stages/stage5/spells/atmospheric_discharge.c:40]: no i misread

File: src/stages/stage5/spells/artificial_lightning.c (hunk 12 above)
Comment (Akaricchi) [reply src/stages/stage5/spells/artificial_lightning.c:87]: Ah yes, you are correct. I missed the context.

The `_i` that the `x` is replacing is the iteration count (i think), not a timer. So it makes no sense to add the wait time to it. It should indeed be `x++, WAIT(3)`, or more "conventionally":
```c
for(int x = 0; x < 67; x++) {
	/* the loop... */
	WAIT(3);
}
```

File: src/stages/stage5/iku.c [hunk 24]
```diff
@@ -14,105 +14,131 @@
...
+	Projectile *p = TASK_BIND(ARGS.p);
+	cmplx m1 = ARGS.m1;
+	cmplx m2 = ARGS.m2;
 
-	float t = time*sqrt(global.diff);
+	for(int time = 0;; time++, YIELD) {
+		int t = time * sqrt(global.diff);
+		if(global.diff > D_Normal && ARGS.mode) {
+			t = time*0.6;
+			t = 230-t;
```
[the remark this answers is not in the document]
Comment (StarWitch) [reply src/stages/stage5/iku.c:126]: What would that look like? Considering `time` is always ticking upwards, when would `time < 0` ever happen?

[the remark this answers is not in the document]
Comment (laochailan) [reply src/stages/stage5/iku.c:126]: if `time*0.6 > 230`, `t < 0` is true. In the original rule this was some kind of time out. Now you could use INVOKE_SUBTASK_DELAYED with the common_enemy_kill (or what it was called) task to do it.

[the remark this answers is not in the document]
Comment (laochailan) [reply src/stages/stage5/iku.c:126]: i mean the projectile equivalent because it is a projectile

File: src/stages/stage5/timeline.c (hunk 17 above)
Comment (StarWitch) [reply src/stages/stage5/timeline.c:39]: There's a few places in the original code where it specifically checks against `D_Hard` and `D_Lunatic` being defined. Example:

```
			PROJECTILE(
				.proto = pp_ball,
				.pos = e->pos,
				.color = RGBA(0.1 + 0.5 * pow((t - 140) / 140.0, 2), 0.0, 0.8, 0.0),
				.rule = accelerated,
				.args = {
					(-2 + (global.diff == D_Hard)) * dir, // <--- this line
					0.02 * dir * (!i || global.diff != D_Lunatic), // <--- also this line
				},
```

It seems a bit silly to me, but I was trying to replicate original behaviour as much as possible.

**Comment by StarWitch** [comment]
Fixed the merge conflict, and rebased it against `remastered-staging`.

File: src/common_tasks.c [hunk 25]
```diff
@@ -241,6 +241,19 @@ int common_charge_custom(
...
+	int delay,
+	int duration,
+	real angle
+) {
+	WAIT(delay);
+	cmplx r = cdir(angle / duration);
+	move->retention *= r;
+	WAIT(duration);
+	move->retention /= r;
+}
```
Comment (laochailan) [inline src/common_tasks.c:255]: `delay` is not needed because `INVOKE_TASK_DELAYED` exists

File: src/stages/stage5/iku.c [hunk 26]
```diff
@@ -14,105 +14,132 @@
...
+
+void iku_lightning_particle(cmplx pos) {
+	char *part = rng_chance(0.5) ? "lightning0" : "lightning1";
+	PARTICLE(
+		.sprite = part,
+		.pos = pos,
+		.color = RGBA(1.0, 1.0, 1.0, 0.0),
+		.timeout = 20,
+		.draw_rule = pdraw_timeout_fade(0.8, 0.0),
+		.angle = rng_real() * M_TAU,
```
Comment (laochailan) [inline src/stages/stage5/iku.c:35]: `rng_angle`

File: src/stages/stage5/nonspells/boss_nonspell_2.c [hunk 27]
```diff
@@ -10,55 +10,58 @@
 
 #include "nonspells.h"
 
-MODERNIZE_THIS_FILE_AND_REMOVE_ME
-
 static cmplx bolts2_laser(Laser *l, float t) {
 	if(t == EVENT_BIRTH) {
+		l->shader = res_shader_optional("lasers/iku_lightning");
```
Comment (laochailan) [inline src/stages/stage5/nonspells/boss_nonspell_2.c:15]: Lasers shaders are obsolete now right? This causes
```
  W: _get_resource: shader program 'lasers/iku_lightning' was not preloaded
102948    D: vfs_set_error: Node 'res/shader/lasers/iku_lightning.prog' does not exist
102948    E: load_shader_program_stage1: VFS error: Node 'res/shader/lasers/iku_lightning.prog' does not exist
```

**Review by laochailan — CHANGES_REQUESTED** [review]
I tried to play the stage but it crashed when the `iku_lightning` shader was loaded and not found.

File: src/stages/stage5/nonspells/boss_nonspell_2.c (hunk 27 above)
Comment (StarWitch) [reply src/stages/stage5/nonspells/boss_nonspell_2.c]: Fixed!

**Review by laochailan — APPROVED** [review]
looking good now. ready for merging?

**Comment by StarWitch** [comment]
Only thing holding back this PR is the `overload` spell. It's currently using the old system because of how galaxy brain it is. It still functions in-situ, of course, and maybe now that we have live-reloading work on it could progress faster.

My own idea is to just get this (and the other stage PRs) merged if the bulk of the coroutine/modernization work is done, so we can start fine-tuning and ironing out wrinkles (like `overload`) instead of feeling like these are huge PRs we have to completely context-change in order to understand fully. I feel like it'd actually get us a v1.4 release this year.

**Comment by laochailan** [comment]
I agree. If it works in situ it should be fine.

**Comment by Akaricchi** [comment]
Merged in 46ad7b183bd5ed97afa45d637ab4b6e2647449f4
